File: Day293/printMatrixInSnakePattern.go
package main

import (
	"fmt"
)

func snakePattern(matrix [][]int) []int {
	var answer []int
	lr := true
	for i := 0; i < len(matrix); i++ {
		if lr {
			for j := 0; j < len(matrix[0]); j++ {
				answer = append(answer, matrix[i][j])
			}
			lr = false
		} else {
			for j := len(matrix[0]) - 1; j >= 0; j-- {
				answer = append(answer, matrix[i][j])
			}
			lr = true
		}
	}
	return answer
}

func main() {
	var t int
	fmt.Scan(&t)

	for t > 0 {
		var n int
		fmt.Scan(&n)

		matrix := make([][]int, n)
		for i := 0; i < n; i++ {
			matrix[i] = make([]int, n)
			for j := 0; j < n; j++ {
				fmt.Scan(&matrix[i][j])
			}
		}

		ans := snakePattern(matrix)
		for _, val := range ans {
			fmt.Print(val, " ")
		}
		fmt.Println()
		t--
	}
}

File: Day293/printMatrixInSnakePattern.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            int n = sc.nextInt();
            int matrix[][] = new int[n][n];
            
            for(int i = 0; i < n; i++)
            {
                for(int j = 0; j < n; j++)
                 matrix[i][j] = sc.nextInt();
            }
            
            Solution ob = new Solution();
            ArrayList<Integer> ans = ob.snakePattern(matrix);
            for (Integer val: ans) 
                System.out.print(val+" "); 
            System.out.println();
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution
{
    //Function to return list of integers visited in snake pattern in matrix.
    static ArrayList<Integer> snakePattern(int matrix[][])
    {
        // code here 
        ArrayList<Integer> answer = new ArrayList<>();
        boolean lr = true;
        for (int i=0; i<matrix.length; i++) {
            if (lr) {
                for (int j=0; j<matrix[0].length; j++) {
                    answer.add(matrix[i][j]);
                }
                lr = false;
            } else {
                for (int j=matrix[0].length-1; j>=0; j--) {
                    answer.add(matrix[i][j]);
                }
                lr = true;
            }
        }
        return answer;
    }
}